vortex-ffi/src/lib.rs

```diff
@@ -80,8 +119,38 @@ pub(crate) unsafe fn to_string_vec(ptr: *const *const c_char, len: c_int) -> Vec
         .collect()
 }
 
-/// Attempt to shutdown the shared tokio runtime if no sessions are active.
-/// May block indefinitely if the runtime is still running tasks.
+/// Attempt to shutdown the shared Tokio runtime if no active sessions exist.
+/// This function is thread-safe & idempotent. It can be called concurrently from multiple threads.
+/// It is recommended to call this function just before exiting the application, after all sessions
+/// have been explicitly dropped via `vx_session_free`.
+///
+/// ## Behavior
+///
+/// - **Success**: If no sessions hold runtime references, shuts down immediately  
+/// - **Failure**: If sessions still exist, runtime continues running (no-op)
+/// - **Blocking**: May block indefinitely if runtime has active tasks
+///
+/// ## Thread Safety
+///
+/// This function is thread-safe but coordination is required for predictable behavior:
+///
+/// - **Race Condition Safe**: Uses atomic Arc operations to detect active references
+/// - **Concurrent Calls**: Multiple threads can call this simultaneously (idempotent)
+/// - **Session Creation Race**: If called during session creation, may fail to shutdown
+///
+/// ## Example Usage
+///
+/// ```c
+/// // C usage pattern
+/// vx_session* session1 = vx_session_new();
+/// vx_session* session2 = vx_session_new();  // Reuses same runtime
+///
+/// vx_session_free(session1);
+/// vx_session_free(session2);
+///
+/// // Explicit cleanup:
+/// vx_try_shutdown_runtime();  // Only succeeds if no sessions active
+/// ```
 #[unsafe(no_mangle)]
 pub extern "C" fn vx_try_shutdown_runtime() {
```

Comment:
I can't imagine we actually want to commit this to as a stable API...?

We should make it clear (in another PR) the state of this API. Honestly, I'd rather just kill it for now and publish language-specific bindings that are allowed to make lock-step breaking changes to their own C APIs.

Reply:
oh yeah, this is gross. it's just that not cleaning up the runtime can lead to segfaults on exit when e.g., used with mimalloc. we could also kill this function and have the runtime be cleaned up when the last VortexSession is dropped

Reply:
Not that I disagree but this is how the api would look like in C anyway? You're sahying we shouldn't have C api and only C++?